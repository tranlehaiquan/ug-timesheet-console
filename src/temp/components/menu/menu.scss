.menu-min-width {
  min-width: 10rem;
}
