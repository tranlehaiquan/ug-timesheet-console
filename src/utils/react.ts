import * as React from 'react'

type IdentityFunction = <T>(fn: T) => T

export const typedMemo: IdentityFunction = React.memo

export const getDisplayName = <T extends any = any>(WrappedComponent: React.ComponentType<T>): string => {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component'
}
