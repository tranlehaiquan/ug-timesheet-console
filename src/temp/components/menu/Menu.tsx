import * as React from "react";
import classnames from "classnames";

export interface IPropsSub extends React.HTMLAttributes<HTMLElement> {
  className?: string;
}

export const Menu: React.FC<IPropsSub> = ({
  children,
  className = "",
  ...props
}) => {
  const ulClasses = "sk-list-none";
  const menuItemClasses =
    "menu-min-width sk-cursor-pointer sk-text-sm sk-text-navy sk-font-normal sk-bg-white sk-max-w-xs sk-mt-1 sk-py-1 sk-border sk-border-solid sk-border-grey-lighter sk-shadow sk-rounded-medium";

  return (
    <ul
      {...props}
      className={classnames(ulClasses, menuItemClasses, className)}
    >
      {children}
    </ul>
  );
};

export const MenuItem: React.FC<IPropsSub> = ({
  children,
  className = "",
  ...props
}) => {
  const liClasses = "sk-py-2 hover:sk-bg-grey-lightest sk-px-3 sk-break-words";
  return (
    <li {...props} className={classnames(liClasses, className)}>
      {children}
    </li>
  );
};
