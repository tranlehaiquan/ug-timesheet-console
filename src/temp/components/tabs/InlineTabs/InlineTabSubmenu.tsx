import * as React from "react";
import classNames from "classnames";

import { Menu, MenuItem } from "../../menu/Menu";

export interface SubmenuItem {
  name: string;
  id: string;
  icon?: string;
  iconClasses?: string;
}

interface IProps extends React.HTMLAttributes<HTMLUListElement> {
  items: SubmenuItem[];
  activeItem: SubmenuItem["id"];
  onItemClick: (itemId: SubmenuItem["id"]) => () => void;
  onMouseLeave: () => void;
}

export const MenuNotification = ({
  text,
  className,
  ...otherProps
}: {
  text: string;
  className: string;
}) => (
  <div
    {...otherProps}
    className={classNames(
      "sk-w-5 sk-h-5 sk-rounded-full sk-bg-blue-lightest sk-inline-block sk-ml-1",
      className
    )}
  >
    <p className="sk-text-blue-600 sk-font-semibold sk-text-xxs sk-m-0 sk-flex sk-justify-center sk-items-center sk-h-full">
      {text}
    </p>
  </div>
);

export const InlineTabSubmenu = ({
  items,
  onMouseLeave,
  activeItem,
  onItemClick,
  ...otherProps
}: IProps) => (
  <Menu onMouseLeave={onMouseLeave}>
    {items.map((item) => (
      <MenuItem
        key={item.id}
        onClick={onItemClick(item.id)}
        className={classNames({
          "sk-text-navy sk-font-semibold": activeItem === item.id,
        })}
      >
        {item.name}
        {item.icon && (
          <MenuNotification
            text={item.icon}
            className={item.iconClasses || ""}
          />
        )}
      </MenuItem>
    ))}
  </Menu>
);
