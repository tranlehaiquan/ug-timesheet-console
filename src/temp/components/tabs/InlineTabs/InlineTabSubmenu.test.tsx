import * as React from 'react'
import { InlineTabSubmenu, MenuNotification } from './InlineTabSubmenu'
import { render, shallow } from 'enzyme'

import { MenuItem } from '../../menu/Menu'

describe('InlineTabSubmenu', () => {
  const items = [{ id: 'hello', name: 'Hello' }, { id: 'world', name: 'World', icon: '2' }]
  const itemClickSpy = jest.fn()

  afterEach(() => {
    itemClickSpy.mockRestore()
  })

  test('renders', () => {
    // Arrange
    const component = render(<InlineTabSubmenu items={ items } onMouseLeave={ jest.fn() } activeItem={ items[0].id } onItemClick={ itemClickSpy } />)

    // Assert
    expect(component).toMatchSnapshot()
  })

  test('renders icon', () => {
    // Arrange
    const component = shallow(<InlineTabSubmenu items={ items } onMouseLeave={ jest.fn() } activeItem={ items[0].id } onItemClick={ itemClickSpy } />)

    // Assert
    expect(component.find(MenuNotification)).toHaveLength(1)
    expect(
      component
        .find(MenuNotification)
        .dive()
        .find('p')
        .first()
        .text()
    ).toBe('2')
  })

  test('click item returns correct payload', () => {
    // Arrange
    const component = shallow(<InlineTabSubmenu items={ items } onMouseLeave={ jest.fn() } activeItem={ items[0].id } onItemClick={ itemClickSpy } />)
    const firstItem = component.find(MenuItem).first()
    const secondItem = component.find(MenuItem).at(1)

    // Act
    firstItem.simulate('click')
    expect(itemClickSpy).toBeCalledWith(items[0].id)

    secondItem.simulate('click')
    expect(itemClickSpy).toBeCalledWith(items[1].id)
  })
})
