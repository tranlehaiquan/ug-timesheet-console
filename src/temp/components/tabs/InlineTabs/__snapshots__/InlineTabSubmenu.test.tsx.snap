// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InlineTabSubmenu renders 1`] = `
<ul
  class="sk-list-none menu-min-width sk-cursor-pointer sk-text-sm sk-text-navy sk-font-normal sk-bg-white sk-max-w-xs sk-mt-1 sk-py-1 sk-border sk-border-solid sk-border-grey-lighter sk-shadow sk-rounded-medium"
>
  <li
    class="sk-py-2 hover:sk-bg-grey-lightest sk-px-3 sk-break-words sk-text-navy sk-font-semibold"
  >
    Hello
  </li>
  <li
    class="sk-py-2 hover:sk-bg-grey-lightest sk-px-3 sk-break-words"
  >
    World
    <div
      class="sk-w-5 sk-h-5 sk-rounded-full sk-bg-blue-lightest sk-inline-block sk-ml-1"
    >
      <p
        class="sk-text-blue-600 sk-font-semibold sk-text-xxs sk-m-0 sk-flex sk-justify-center sk-items-center sk-h-full"
      >
        2
      </p>
    </div>
  </li>
</ul>
`;
